pkgs/top-level/module/tests.nix

```diff
@@ -0,0 +1,115 @@
+/*
+  Run
+  [nixpkgs]$ nix-build -A tests.top-level.module --show-trace
+
+  Repl
+  [nixpkgs]$ nix repl pkgs/top-level/module/test.nix --show-trace
+  nix-repl> tests
+
+  Maintainer: @roberth
+*/
+
+# A the tests are invoked using an already invoked nixpkgs, we're going to
+# mostly ignore this pkgs.
+# To catch silly mistakes, let's rename this to pkgs_.
+{ pkgs }:
+let pkgs_ = pkgs;
+in
+let pkgs = throw "You probably didn't mean to use pkgs, but rather use a more specific pkgs variable in this test.";
+
+    inherit (pkgs_) lib emptyFile;
+    inherit (lib) evalModules;
+in
+
+# We define use a rec {} mostly as a let binding, as the tests attribute is all
+# that really matters for running the tests. However, having access to all the
+# other attributes is useful for debugging.
+rec {
+
+  # Helper functions
+  noValues = lib.mapAttrs (_: _: null);
+  shouldBe = actual: expected: if actual == expected then true else
+    builtins.trace "actual:"
+    builtins.trace actual
+    builtins.trace "expected:"
+    builtins.trace expected
+    false;
+
+  invokeModule = callerModule: evalModules { modules = [ ./module.nix callerModule ]; };
+
+  basic = invokeModule { hostPlatform = "x86_64-linux"; };
+  basicExpect = import ../default.nix { localSystem = "x86_64-linux"; };
+
+  cross = invokeModule { buildPlatform = "x86_64-linux"; hostPlatform = "aarch64-linux"; };
+  crossExpect = import ../default.nix { localSystem = "x86_64-linux"; crossSystem = "aarch64-linux"; };
+
+  overlayExample = self: super: { hello = super.hello.overrideAttrs (old: { name = "hello2"; }); };
+
+  overlay = invokeModule {
+    hostPlatform = "x86_64-linux";
+    overlays = [ overlayExample ];
+  };
+  # overlayExpect = import ../default.nix { localSystem = "x86_64-linux"; overlays = [ overlayExample ]; };
+
+  havingConfig = invokeModule {
+    # Defining a value for an option named `config` is a bit awkward withou
+    # the shorthand syntax. This is noted in the module file's top comment.
+    config = {
+      hostPlatform = "x86_64-linux";
+      config = {
+        # Not the most popular config option, but one that is easy to test.
+        perlPackageOverrides = pkgs: { just-a-proof-of-use = true; };
+      };
+    };
+  };
+
+  optimizedSimple = invokeModule {
+    hostPlatform = "x86_64-linux";
+    _optimize = args@{ pkgs, inputsSet, ... }:
+      pkgs // { testData = args; };
+  };
+
+  optimizedAllOpts = invokeModule {
+    config = {
+      buildPlatform = "x86_64-linux";
+      hostPlatform = "aarch64-linux";
+      config = { allowAliases = false; allowUnfree = true; };
+      overlays = [ overlayExample ];
+      _optimize = args@{ pkgs, inputsSet, ... }:
+        pkgs // { testData = args; };
+    };
+  };
+  allInputsSet = {
```

Comment:
```suggestion

  # All inputs that effect the construction of the package set / arguments to _optimize.
  # Not `_optimize` because it is not an argument to itself.
  # Not `pkgs` because it is not an argument and the result of the `_optimize` call.
  allInputsSet = {
```